torch/csrc/jit/script/compiler.cpp

```diff
@@ -2451,31 +2341,49 @@ struct to_ir {
     return graph->insertConstant(stack[0], nullptr, tree->range());
   }
 
-  // This function extract a new graph from its original subgraph
   std::shared_ptr<SugaredValue> emitForkExpr(
       SourceRange loc,
       const std::shared_ptr<SugaredValue>& forked,
       at::ArrayRef<NamedValue> inputs,
       at::ArrayRef<NamedValue> attributes) {
-    // Build the fork node without inputs
-    auto fork_node = method.graph()
-                         ->insertNode(method.graph()->create(prim::fork, 1))
-                         ->setSourceRange(loc);
-    auto body_block = fork_node->addBlock();
-
-    // Build a template of the graph to be executed
-    Value* node_output;
-    {
-      WithInsertPoint guard(body_block);
-      auto fn_sugared_output = forked->call(loc, method, inputs, attributes, 1);
-      auto fn_simple_output = fn_sugared_output->asValue(loc, method);
-      body_block->registerOutput(fn_simple_output);
-      node_output = fork_node->output()->setType(
-          FutureType::create(fn_simple_output->type()));
-    }
-    // Lambda lift block(0) into attr::Subgraph
-    lambdaLiftFork(fork_node);
-    runCleanupPasses(fork_node->g(attr::Subgraph));
+    auto g = method.graph();
+    Node* fork_node;
+    TypePtr out_type;
+
+    // We have two pathways for forks, one for invocations of closures
+    // and one for invocations of builtins or functions. If the input value is
+    // simple than we go through the closure pathway.
+    auto simple_forked = asSimple(forked);
+    if (simple_forked) {
+      // because closures are not first-class yet, they are represented as
+      // a tuple of (Output Type, Context). We need to extract the output type
+      // so the rest of the function type checks. Later, we ensure that
+      // the input actually is a closure and not just a tuple, and inline the
+      // closure into the fork.
+      auto tuple_type = simple_forked->type()->cast<TupleType>();
+      if (!tuple_type) {
+        throw ErrorReport(loc) << "Cannot fork this value";
+      }
+      fork_node =
```

Comment:
This seems really redundant:
1. we are creating a 'forkClosure', different than a normal closure.
2. We also still have the other fork node!
3. we do a lot of work to deduplicate this!

Why not make prim::fork be : `prim::fork(the_closure_to_fork)`. If we already have a closure then this is as simply as just generating the prim::fork call and setting the type. If it don't have a closure, we _create one here_  using the code below which already is making a sub-block closure thing.

This way a lot of the code in the cleanup passes can just be deleted. Only closure exists as a sub-block thing to handle ssa conversion on. All fork nodes can be trivially lowered by taking the lambda lifted closure and dropping its function into the subgraph of the closure and its closed arguments into  the fork command. It is also much closer to how fork will work once we really do have closures.

Reply:
I think creating a Closure sugared value will solve these problems, as is, there needs to be two sets of logic because the closure simple value is not callable.